src/daemon/rpmostreed-transaction-types.cxx

```diff
@@ -787,6 +791,117 @@ deploy_has_bool_option (DeployTransaction *self,
   return vardict_lookup_bool (self->options, option, FALSE);
 }
 
+static const char *
+deploy_has_string_option (DeployTransaction *self,
+                          const char        *option)
+{
+  return static_cast<const char*>(vardict_lookup_ptr (self->options, option, "s"));
+}
+
+/* Write a state file which records information about the agent that is "driving" updates */
+static gboolean
+record_driver_info (RpmostreedTransaction *transaction,
+                    const gchar           *update_driver,
+                    GCancellable          *cancellable,
+                    GError               **error)
+{
+  g_auto(GVariantBuilder) caller_info_builder;
+  g_variant_builder_init (&caller_info_builder, G_VARIANT_TYPE ("a{sv}"));
+
+  const char *sd_unit = rpmostreed_transaction_get_sd_unit (transaction);
+  if (!update_driver)
+    {
+      g_set_error (error, RPM_OSTREED_ERROR,
+                   RPM_OSTREED_ERROR_FAILED, "update driver name not provided");
+      return FALSE;
+    }
+  g_variant_builder_add (&caller_info_builder, "{sv}", RPMOSTREE_DRIVER_NAME,
+                         g_variant_new_string (update_driver));
+  if (!sd_unit)
+    {
+      g_set_error (error, RPM_OSTREED_ERROR,
+                   RPM_OSTREED_ERROR_FAILED, "could not find caller systemd unit");
+      return FALSE;
+    }
+  g_variant_builder_add (&caller_info_builder, "{sv}", RPMOSTREE_DRIVER_SD_UNIT,
+                         g_variant_new_string (sd_unit));
+
+  g_autoptr(GVariant) driver_info =
+    g_variant_ref_sink (g_variant_builder_end (&caller_info_builder));
+
+  if (!glnx_shutil_mkdir_p_at (AT_FDCWD, RPMOSTREE_RUN_DIR, 0755, cancellable, error))
+    return FALSE;
+  if (!glnx_file_replace_contents_at (AT_FDCWD, RPMOSTREE_DRIVER_STATE,
+                                      static_cast<const guint8*>(g_variant_get_data (driver_info)),
+                                      g_variant_get_size (driver_info),
+                                      static_cast<GLnxFileReplaceFlags>(0), cancellable, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+gboolean
+get_driver_g_variant (GVariant **driver_g_variant,
+                      GError   **error)
+{
+  glnx_autofd int fd = -1;
+  g_autoptr(GError) local_error = NULL;
+  if (!glnx_openat_rdonly (AT_FDCWD, RPMOSTREE_DRIVER_STATE, TRUE, &fd,
+                           &local_error))
+    {
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        return g_propagate_error (error, util::move_nullify (local_error)), FALSE;
+      return TRUE;
+    }
+
+  g_autoptr(GBytes) data = glnx_fd_readall_bytes (fd, NULL, error);
+  if (!data)
+    return FALSE;
+
+  *driver_g_variant =
+    g_variant_ref_sink (g_variant_new_from_bytes (G_VARIANT_TYPE_VARDICT, data, FALSE));
```

Comment:
It'd be a bit more idiomatic to have the return value here be `GVariant*` and `NULL` on error.